mesonbuild/build.py

```diff
@@ -1191,12 +1191,22 @@ def process_kwargs(self, kwargs, environment):
         # Visual Studio module-definitions file
         if 'vs_module_defs' in kwargs:
             path = kwargs['vs_module_defs']
-            if os.path.isabs(path):
-                self.vs_module_defs = File.from_absolute_file(path)
+            if isinstance(path, str):
+                if os.path.isabs(path):
+                    self.vs_module_defs = File.from_absolute_file(path)
+                else:
+                    self.vs_module_defs = File.from_source_file(environment.source_dir, self.subdir, path)
+                # link_depends can be an absolute path or relative to self.subdir
+                self.link_depends.append(path)
+            elif isinstance(path, File):
+                # When passing a generated file.
+                self.vs_module_defs = path
+                # link_depends can be an absolute path or relative to self.subdir
+                self.link_depends.append(path.absolute_path(environment.source_dir, environment.build_dir))
             else:
-                self.vs_module_defs = File.from_source_file(environment.source_dir, self.subdir, path)
-            # link_depends can be an absolute path or relative to self.subdir
-            self.link_depends.append(path)
+                raise InvalidArguments(
+                    'Shared library vs_module_defs must be either a string, '
+                    'list, or a file object')
```

Comment:
nope, rebase error there. I started to add support for empty or one element lists, and then decided that wasn't really a good idea.